src/sql/src/pure.rs

```diff
@@ -384,7 +385,213 @@ pub async fn purify_create_source(
+
+                        let schemas = match tables_by_name.get(&upstream_name.item) {
+                            Some(schemas) => schemas,
+                            None => bail!("table {upstream_name} not found in source"),
+                        };
+
+                        let schema = match &upstream_name.schema {
+                            Some(schema) => schema,
+                            None => match schemas.keys().exactly_one() {
+                                Ok(schema) => schema,
+                                Err(_) => bail!("table {upstream_name} is ambiguous, consider specifying the schema"),
+                            },
+                        };
+
+                        let databases = match schemas.get(schema) {
+                            Some(databases) => databases,
+                            None => bail!("schema {schema} not found in source"),
+                        };
+
+                        let database = match &upstream_name.database {
+                            Some(database) => database,
+                            None => match databases.keys().exactly_one() {
+                                Ok(database) => database,
+                                Err(_) => bail!("table {upstream_name} is ambiguous, consider specifying the database"),
+                            },
+                        };
+
+                        let desc = match databases.get(database) {
+                            Some(desc) => desc.clone(),
+                            None => bail!("database {database} not found source"),
+                        };
+
+                        let qualified_upstream_name = UnresolvedObjectName::qualified(&[
+                            database,
+                            schema,
+                            &upstream_name.item,
+                        ]);
+                        validated_requested_subsources.push((
+                            qualified_upstream_name,
+                            subsource_name,
+                            desc,
+                        ));
+                    }
+                }
+                None => {
+                    if available_subsources.is_none() {
+                        bail!("multi-output sources require a FOR TABLES (..) or FOR ALL TABLES statement");
+                    }
+                }
+            };
+
+            // Now that we have an explicit list of validated requested subsources we can create them
+            for (i, (upstream_name, subsource_name, desc)) in
+                validated_requested_subsources.into_iter().enumerate()
+            {
+                // Figure out the schema of the subsource
+                let mut columns = vec![];
+                for (column_name, column_type) in desc.iter() {
+                    let name = Ident::new(column_name.as_str().to_owned());
+
+                    // TODO(petrosagg): This is gross but unclear if there is a better way to go
```

Comment:
You can shorten this nominally  by using `sql::plan::statement::StatementContext::resolve_type`, and taking `columntype.scalar_type` into a `pgrepr::types::Type` using `from`/`into`. Maybe you still want to prevent other types from getting used, so not much shorter, but is at least well structured.

Reply:
I did not see that ScalarType had an `Into<Type>` implementation, this is so much nicer, thanks!